[Autograd] Improve error for leaf tensors as out argument to fallback

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121089
* #121142

Closes  #120988

Currently operators that hit the autograd fallback call `check_inplace`
on all mutated inputs, including out arguments. This leads to a slightly
confusing error message:
```
RuntimeError: a leaf Variable that requires grad is being used in an in-place operation.
```

Compared to functions that don't fallback, which raise
```
RuntimeError: add(): functions with out=... arguments don't support automatic differentiation, but one of the arguments requires grad.
```

This changes the error message to make clear the issue is with the out argument,
but does not tighten the check to outright ban out arguments that require grad.
Instead, I use the same checks from `check_inplace` which allows non-leaf tensors
that require grad to pass without error.

peterbell10 committed:
[Autograd] Improve error for leaf tensors as out argument to fallback

Currently operators that hit the autograd fallback call `check_inplace`
on all mutated inputs, including out arguments. This leads to a slightly
confusing error message:
```python
RuntimeError: a leaf Variable that requires grad is being used in an in-place operation.
```

Compared to functions that don't fallback, which raise
```python
RuntimeError: add(): functions with out=... arguments don't support automatic differentiation, but one of the arguments requires grad.
``

This changes the error message to make clear the issue is with the out argument,
but does not tighten the check to outright ban out arguments that require grad.
Instead, I use the same checks from `check_inplace` which allows non-leaf tensors
that require grad to pass without error.

[ghstack-poisoned]

peterbell10 committed:
Update on "[Autograd] Improve error for leaf tensors as out argument to fallback"


Closes  #120988

Currently operators that hit the autograd fallback call `check_inplace`
on all mutated inputs, including out arguments. This leads to a slightly
confusing error message:
```python
RuntimeError: a leaf Variable that requires grad is being used in an in-place operation.
```

Compared to functions that don't fallback, which raise
```python
RuntimeError: add(): functions with out=... arguments don't support automatic differentiation, but one of the arguments requires grad.
``

This changes the error message to make clear the issue is with the out argument,
but does not tighten the check to outright ban out arguments that require grad.
Instead, I use the same checks from `check_inplace` which allows non-leaf tensors
that require grad to pass without error.

[ghstack-poisoned]

peterbell10 committed:
Update on "[Autograd] Improve error for leaf tensors as out argument to fallback"


Closes  #120988

Currently operators that hit the autograd fallback call `check_inplace`
on all mutated inputs, including out arguments. This leads to a slightly
confusing error message:
```
RuntimeError: a leaf Variable that requires grad is being used in an in-place operation.
```

Compared to functions that don't fallback, which raise
```
RuntimeError: add(): functions with out=... arguments don't support automatic differentiation, but one of the arguments requires grad.
```

This changes the error message to make clear the issue is with the out argument,
but does not tighten the check to outright ban out arguments that require grad.
Instead, I use the same checks from `check_inplace` which allows non-leaf tensors
that require grad to pass without error.

[ghstack-poisoned]

lezcano commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -58,6 +58,26 @@ inline void check_inplace(const at::Tensor& tensor, bool requires_grad) {
 }
 
+// Same as check_inplace but returns false instead of raising an error
+inline bool can_mutate_inplace(const at::Tensor& tensor, bool requires_grad) {
```
You could make a `std::optional<std::string> _can_mutate_inplace` to avoid duplicating the logic.

peterbell10 replied:
`handle_view_on_rebase` can throw many different errors or none at all, so I don't think it's worth it. At that point you'd be as well doing `try-catch` around `check_inplace`.

soulitzer replied:
hmmm still feels bad to duplicate imo

The code change doesn't seem too bad since there's only a single `TORCH_CHECK` call we need to convert to a `return msg` at the very end, but maybe I'm missing something?

The comment above the call to `handle_view_on_rebase` in `check_inplace` is out of date btw, we don't need to worry about warns anymore.

lezcano commented on torch/csrc/autograd/autograd_not_implemented_fallback.cpp:
```diff
@@ -301,11 +301,27 @@ static void autogradNotImplementedFallbackImpl(
 
   const bool any_requires_grad = !tensors_requiring_grad_on_stack.empty();
+  bool has_out_arg = false;
+  for (const auto& arg : schema.arguments()) {
```
nit now we have `std::any`

soulitzer reviewed (commented):
Thanks for the fix! Still a little concerned about the logic duplication though

peterbell10 committed:
Use any_of on "[Autograd] Improve error for leaf tensors as out argument to fallback"


Closes  #120988

Currently operators that hit the autograd fallback call `check_inplace`
on all mutated inputs, including out arguments. This leads to a slightly
confusing error message:
```
RuntimeError: a leaf Variable that requires grad is being used in an in-place operation.
```

Compared to functions that don't fallback, which raise
```
RuntimeError: add(): functions with out=... arguments don't support automatic differentiation, but one of the arguments requires grad.
```

This changes the error message to make clear the issue is with the out argument,
but does not tighten the check to outright ban out arguments that require grad.
Instead, I use the same checks from `check_inplace` which allows non-leaf tensors
that require grad to pass without error.

[ghstack-poisoned]

peterbell10 committed:
Reduce logic duplication on "[Autograd] Improve error for leaf tensors as out argument to fallback"


Closes  #120988

Currently operators that hit the autograd fallback call `check_inplace`
on all mutated inputs, including out arguments. This leads to a slightly
confusing error message:
```
RuntimeError: a leaf Variable that requires grad is being used in an in-place operation.
```

Compared to functions that don't fallback, which raise
```
RuntimeError: add(): functions with out=... arguments don't support automatic differentiation, but one of the arguments requires grad.
```

This changes the error message to make clear the issue is with the out argument,
but does not tighten the check to outright ban out arguments that require grad.
Instead, I use the same checks from `check_inplace` which allows non-leaf tensors
that require grad to pass without error.

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge
